executor/insert_common.go

```diff
@@ -258,18 +264,31 @@ func (e *InsertValues) insertRowsFromSelect(ctx context.Context, cols []*table.C
 		}
 
 		for innerChunkRow := iter.Begin(); innerChunkRow != iter.End(); innerChunkRow = iter.Next() {
-			innerRow := innerChunkRow.GetDatumRow(fields)
+			innerRow := types.CopyRow(innerChunkRow.GetDatumRow(fields))
 			e.rowCount++
 			row, err := e.fillRowData(cols, innerRow)
 			if err != nil {
 				return errors.Trace(err)
 			}
 			rows = append(rows, row)
+			if batchInsert && e.rowCount%uint64(batchSize) == 0 {
+				if err := exec(rows); err != nil {
+					return errors.Trace(err)
+				}
+				e.ctx.StmtCommit()
```

Comment:
This operation will leave dirty data to binlog if error happen later.
We should never call `StmtCommit` before a statement is finished with no error.